pl_examples/loop_examples/kfold.py

```diff
@@ -197,7 +205,8 @@ def on_advance_end(self) -> None:
         self.trainer.save_checkpoint(osp.join(self.export_path, f"model.{self.current_fold}.pt"))
         # restore the original weights + optimizers and schedulers.
         self.trainer.lightning_module.load_state_dict(self.lightning_module_state_dict)
-        self.trainer.accelerator.setup_optimizers(self.trainer)
+        self.trainer.strategy.setup_optimizers(self.trainer)
+        self.replace(fit_loop=FitLoop(self.fit_loop.min_epochs, self.fit_loop.max_epochs))
```

Comment:
here we are resetting the interval fit loop progress to make sure that at every next fold it starts from the 0th epoch.

Reply:
I tried your change on colab file and it raise:

```
<ipython-input-53-c465e20ce3ca> in run()
     17     trainer.fit_loop.connect(internal_fit_loop)
     18 
---> 19     trainer.fit(model, datamodule=datamodule)
     20     trainer.test(model, datamodule=datamodule)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloaders, val_dataloaders, datamodule, train_dataloader, ckpt_path)
    736             train_dataloaders = train_dataloader
    737         self._call_and_handle_interrupt(
--> 738             self._fit_impl, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path
    739         )
    740 

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _call_and_handle_interrupt(self, trainer_fn, *args, **kwargs)
    680         """
    681         try:
--> 682             return trainer_fn(*args, **kwargs)
    683         # TODO: treat KeyboardInterrupt as BaseException (delete the code below) in v1.7
    684         except KeyboardInterrupt as exception:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _fit_impl(self, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path)
    770         # TODO: ckpt_path only in v1.7
    771         ckpt_path = ckpt_path or self.resume_from_checkpoint
--> 772         self._run(model, ckpt_path=ckpt_path)
    773 
    774         assert self.state.stopped

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _run(self, model, ckpt_path)
   1192 
   1193         # dispatch `start_training` or `start_evaluating` or `start_predicting`
-> 1194         self._dispatch()
   1195 
   1196         # plugin will finalized fitting (e.g. ddp_spawn will load trained model)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _dispatch(self)
   1272             self.training_type_plugin.start_predicting(self)
   1273         else:
-> 1274             self.training_type_plugin.start_training(self)
   1275 
   1276     def run_stage(self):

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py in start_training(self, trainer)
    200     def start_training(self, trainer: "pl.Trainer") -> None:
    201         # double dispatch to initiate the training loop
--> 202         self._results = trainer.run_stage()
    203 
    204     def start_evaluating(self, trainer: "pl.Trainer") -> None:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in run_stage(self)
   1282         if self.predicting:
   1283             return self._run_predict()
-> 1284         return self._run_train()
   1285 
   1286     def _pre_training_routine(self):

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _run_train(self)
   1312         self.fit_loop.trainer = self
   1313         with torch.autograd.set_detect_anomaly(self._detect_anomaly):
-> 1314             self.fit_loop.run()
   1315 
   1316     def _run_evaluate(self) -> _EVALUATE_OUTPUT:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/loops/base.py in run(self, *args, **kwargs)
    144                 self.on_advance_start(*args, **kwargs)
    145                 self.advance(*args, **kwargs)
--> 146                 self.on_advance_end()
    147                 self.restarting = False
    148             except StopIteration:

<ipython-input-52-ddd2156c6f38> in on_advance_end(self)
    169         # restore the original weights + optimizers and schedulers.
    170         self.trainer.lightning_module.load_state_dict(self.lightning_module_state_dict)
--> 171         self.trainer.strategy.setup_optimizers(self.trainer)
    172         self.replace(fit_loop=FitLoop(self.fit_loop.min_epochs, self.fit_loop.max_epochs))
    173 

AttributeError: 'Trainer' object has no attribute 'strategy'
```

pytorch lightning version==1.5.6 and all envs are same as fisrt.

Reply:
If I remove line 209,
`self.trainer.strategy.setup_optimizers(self.trainer)`

it raise:
```python
<ipython-input-16-c465e20ce3ca> in run()
     17     trainer.fit_loop.connect(internal_fit_loop)
     18 
---> 19     trainer.fit(model, datamodule=datamodule)
     20     trainer.test(model, datamodule=datamodule)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloaders, val_dataloaders, datamodule, train_dataloader, ckpt_path)
    736             train_dataloaders = train_dataloader
    737         self._call_and_handle_interrupt(
--> 738             self._fit_impl, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path
    739         )
    740 

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _call_and_handle_interrupt(self, trainer_fn, *args, **kwargs)
    680         """
    681         try:
--> 682             return trainer_fn(*args, **kwargs)
    683         # TODO: treat KeyboardInterrupt as BaseException (delete the code below) in v1.7
    684         except KeyboardInterrupt as exception:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _fit_impl(self, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path)
    770         # TODO: ckpt_path only in v1.7
    771         ckpt_path = ckpt_path or self.resume_from_checkpoint
--> 772         self._run(model, ckpt_path=ckpt_path)
    773 
    774         assert self.state.stopped

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _run(self, model, ckpt_path)
   1192 
   1193         # dispatch `start_training` or `start_evaluating` or `start_predicting`
-> 1194         self._dispatch()
   1195 
   1196         # plugin will finalized fitting (e.g. ddp_spawn will load trained model)

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _dispatch(self)
   1272             self.training_type_plugin.start_predicting(self)
   1273         else:
-> 1274             self.training_type_plugin.start_training(self)
   1275 
   1276     def run_stage(self):

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py in start_training(self, trainer)
    200     def start_training(self, trainer: "pl.Trainer") -> None:
    201         # double dispatch to initiate the training loop
--> 202         self._results = trainer.run_stage()
    203 
    204     def start_evaluating(self, trainer: "pl.Trainer") -> None:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in run_stage(self)
   1282         if self.predicting:
   1283             return self._run_predict()
-> 1284         return self._run_train()
   1285 
   1286     def _pre_training_routine(self):

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py in _run_train(self)
   1312         self.fit_loop.trainer = self
   1313         with torch.autograd.set_detect_anomaly(self._detect_anomaly):
-> 1314             self.fit_loop.run()
   1315 
   1316     def _run_evaluate(self) -> _EVALUATE_OUTPUT:

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/loops/base.py in run(self, *args, **kwargs)
    144                 self.on_advance_start(*args, **kwargs)
    145                 self.advance(*args, **kwargs)
--> 146                 self.on_advance_end()
    147                 self.restarting = False
    148             except StopIteration:

<ipython-input-15-e03269e2a105> in on_advance_end(self)
    171         self.trainer.accelerator.setup_optimizers(self.trainer)
    172         # self.trainer.strategy.setup_optimizers(self.trainer)
--> 173         self.replace(fit_loop=FitLoop(self.fit_loop.min_epochs, self.fit_loop.max_epochs))
    174 
    175     def on_run_end(self) -> None:

<ipython-input-15-e03269e2a105> in __getattr__(self, key)
    204         # requires to be overridden as attributes of the wrapped loop are being accessed.
    205         if key not in self.__dict__:
--> 206             return getattr(self.fit_loop, key)
    207         return self.__dict__[key]

AttributeError: 'FitLoop' object has no attribute 'replace'
```

Reply:
The changes in this PR require Lightning to be installed from master (or nightly)

Reply:
@carmocca do you think the pr should be updated to work with patch releases or master?

Reply:
I'd say master. Loop customization is still experimental

Reply:
Ok I'll try nightly later.
Is there no way to run it on 1.5.6 btw?

Reply:
yes, but few things you might need to change.
```
trainer.strategy.setup_optimizers -> trainer.accelerator.setup_optimizers
fit_loop.replace -> reinitialize fitloop using the same min and max_epochs
```
then I think it should work.

Reply:
Hi I installed from master, but another error occurred. It says that my DataModule doesn't have attribute prepare_data_per_node.

Reply:
> 

Could you be more clear on how to reinitialize fitloop using the same min and max_epochs? Thanks!